### PR TITLE
time-limited shard synchronization

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/16124

* Added startup option `--cluster.shard-synchronization-attempt-timeout` to limit the amount of time to spend in shard synchronization attempts. The default timeout value is 20 minutes.
  Running into the timeout will not lead to a synchronization failure, but will continue the synchronization shortly after. Setting a timeout can help to split the synchronization of large shards into smaller chunks and release snapshots and archived WAL files on the leader earlier.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.9: this PR
  - [ ] Backport for 3.8: -
  - [ ] Backport for 3.7: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by neunhoef — APPROVED** [review]
LGTM. I have tried manually that the change has the desired effect.

File: arangod/Cluster/ReplicationTimeoutFeature.cpp [hunk 1]
```diff
@@ -59,36 +59,62 @@ void ReplicationTimeoutFeature::collectOptions(
...
+          arangodb::options::Flags::Hidden,
+          arangodb::options::Flags::OnDBServer));
+
+  options
+      ->addOption(
+          "--cluster.shard-synchronization-attempt-timeout",
+          "timeout (in seconds) for every shard synchronization attempt. "
+          "running into the timeout will not lead to a synchronization "
+          "failure, but will continue the synchronization shortly after. "
+          "setting a timeout can help to split the replication of large "
```
Comment (cpjulia) [inline arangod/Cluster/ReplicationTimeoutFeature.cpp:100]:
```suggestion
          "Setting a timeout can help to split the replication of large "
```

**Review by cpjulia — APPROVED** [review]
LGTM

File: arangod/Cluster/SynchronizeShard.cpp [hunk 2]
```diff
@@ -604,9 +606,25 @@ static arangodb::ResultT<SyncerId> replicationSynchronize(
 
   auto& agencyCache =
       job.feature().server().getFeature<ClusterFeature>().agencyCache();
+  ReplicationTimeoutFeature& timeouts =
+      job.feature().server().getFeature<ReplicationTimeoutFeature>();
 
-  syncer->setCancellationCheckCallback([=, &agencyCache]() -> bool {
+  syncer->setCancellationCheckCallback([=, &agencyCache, &timeouts]() -> bool {
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Cluster/SynchronizeShard.cpp:612]: It was used before and is still used:
https://github.com/arangodb/arangodb/pull/16144/files/3f09e51f587caa508db62ccc557fbbb6024d5286#diff-e4dc823dbfb17979dae34c6df2abac250095a77c64bb4567989b13a8d1ca04c0R632

**Pull request merged**